packages/react-native/ReactCommon/react/performance/timeline/PerformanceObserverRegistry.h

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <set>
+#include "PerformanceObserver.h"
+
+namespace facebook::react {
+
+class PerformanceObserverRegistry {
+ public:
+  PerformanceObserverRegistry() = default;
+
+  void addObserver(const std::weak_ptr<PerformanceObserver>& observer);
+  void removeObserver(const std::shared_ptr<PerformanceObserver> observer);
+  void removeObserver(const PerformanceObserver& observer);
+
+  void emit(const PerformanceEntry& entry);
+
+ private:
+  mutable std::mutex observersMutex_;
+  std::set<std::weak_ptr<PerformanceObserver>, std::owner_less<std::weak_ptr<PerformanceObserver>>> observers_;
```

Comment:
Given that observers can only be created and destroyed from JS, we could make the ownership of the observers of the registry, and make the reference in the native state a regular pointer to the object owned by the registry. I think that would simplify the model a bit.

Reply:
Sure, if we don't want to support creating observers native-side at all, then I can do it :)

Reply:
My initial goal was to make the registry instance owned by `ReactInstance`, WDYT?

Reply:
Keeping the option to create observers from native is good, so I like the idea of making the registry owned by the `ReactInstance`.